include/envoy/grpc/async_client.h

```diff
@@ -41,7 +42,16 @@ class AsyncStream {
    *                   object, but callbacks may still be received until the stream is closed
    *                   remotely.
    */
-  virtual void sendMessage(const Protobuf::Message& request, bool end_stream) PURE;
+  virtual void sendMessage(const Protobuf::Message& request, bool end_stream);
+
+  /**
+   * Send request message to the stream.
+   * @param request serializalized message.
+   * @param end_stream close the stream locally. No further methods may be invoked on the stream
+   *                   object, but callbacks may still be received until the stream is closed
+   *                   remotely.
+   */
+  virtual void sendRawMessage(Buffer::InstancePtr request, bool end_stream) PURE;
```

Comment:
does the first parameter has to be an unique_ptr? seems `Buffer::Instance&` should be fine?

Reply:
The caller has no further need for the buffer and the callee may need to keep it (queue it) so it makes sense to eliminate the overhead of copying it and pass it as a unique ptr.   In other bits of the code this is done via && (e.g. Buffer::InstancePtr &&), but given that a std::unique_ptr is only 16 bytes that doesn't make much sense to me.

Reply:
This is a generic interface, so it seems like "const Buffer::Instance&" makes the most sense if the copy is reasonably efficient.   What do you guys think?

Reply:
Slice movement is relatively efficient, but that's not copying. It doesn't have to be `const`, you can donate slices between two mutable buffers.

Reply:
Ended up with Buffer::InstancePtr&& which seemed to be most consistent with other interfaces.